오픈 마켓[Step2] - Jost, Jamking

jaemuYeo opened the pull request:
@SungPyo 
안녕하세요 날씨 😀
Step1 리뷰 이후 조금은 늦었지만 약속의 화요일을 맞춰 Step2 PR 보내요!! ㅎㅎ
너무 많은 내용이 있었어서 정신없이 진행했습니다 ,, ><
이번 리뷰도 잘 부탁드려요!!
언제든 대기 중이니 호출해주시면 회의실에서 만나요 😁

### 지난 피드백 반영 내용
- Request 모델 타입을 메서드 별로 분리를 하였습니다.
- Mock 데이터 및 타입에 대한 테스트파일을 Test-Target으로 분리하였습니다.

### 조언받고 싶은 부분 / 고민한점
1. 매직넘버 관련
      - 몇몇 상수로 저장 후 대입한 값에대해 오히려 가독성이 떨어지는 듯 해보입니다. 날씨의 의견이 궁금합니다.
      - calculateLayoutHeight() 에서 3과 5는 상수로 어떻게 표현해야 할지 모르겠습니다. 위의 내용과 비슷한 맥락입니다.
      - 셀이나 특정 뷰의 사이즈를 구하는 계산식의 값들도 매직넘버를 사용하지 말아야하는지 궁금합니다.

2. FlowLayout을 통한 레이아웃 적용 문제
      - Cell에 대한 가로 방향에 대한 문제는 가로 비율에 대응하여 모든 디바이스에 잘 적용되게 해두었습니다.
         다만 세로에 대한 부분에서 길이를 어떤식으로 계산해야 할지 모르겠습니다.
        셀 내부의 UI 요소의 개수나 다른 내용이 많아지는 것에 대비하여 날씨는 어떻게 대응하는지 조언받고 싶습니다!!
        getItemCardHeight(with:) 메서드에서 현재 `(너비 / 3) * 5`로 세로 길이를 적용했는데
        기존에 눈대중으로 그냥 적용했던 180 * 300 에 대한 값을 최대한 비슷한 길이에 맞춰서 모든 디바이스에 맞도록 했습니다.
        사이즈에 대한 부분에 조언 부탁드려요 
        글 적다보니 생각이 들었는데 현업에서는 디자이너가 제공하는 디자인에 따라 비율이 정해져 있을 것이라 생각이 드는데
        실제 현업 프로젝트에서는 디자이너가 정해놓은 비율에 맞춰서 진행하면 되는것일까요??

### 최적화 관련
- 이미지 캐싱을 적용하였습니다. 스크롤하는 과정에서 버벅거림이 있었는데 이 원인이 이미지의 로딩 시간 관련된
일이라고 생각해서 이 부분을 잡아나가게 되었습니다. 기존에는 매 상품마다 썸네일 url을 직접 데이터로 바인딩하여
이미지에 다시 바인딩하는 형태였다면, 이후에는 `ImageManager` 라는 타입을 싱글톤으로 설계하여
`[String : Data]`의 딕셔너리 타입을 통해 Memoization을 적용하였습니다.
PageItemList Data를 fech 시 이미지 rul들에 대해 비동기로 데이터를 미리 생성하는 방법으로 구현하였습니다.
이로 인해 이전에 말씀해주셨던 NSCache를 적용해보지 못하였고 실제로 시뮬레이터(디바이스 포함) 테스트 결과
스크롤이 한층 더 부드러워지는 현상을 경험하였습니다.
이러한 방법으로 최적화 한 것이 괜찮은 방법인지, 아니면 NSCache를 적용하는 방법이 더 적절한 방법인지,
또는 더 최적화 할만한 요소들이 있는지에 대해 조언받고 싶습니다!!

- 이미지 캐싱 적용 후 불러온 데이터들에 대한 부분은 부드럽게 스크롤이 되었지만
다음 페이지에 대한 데이터를 추가로 불러오는 과정에 순간적인 버벅임이 조금은 존재했었습니다.
이 부분을 개선하고자 nextItemList 라는 변수에 다음페이지에 대한 데이터를 미리 담아두고,
다음 페이지에 대한 데이터를 요청할 타이밍에 준비해 둔 데이터를 기존 데이터와 합쳐서 리로딩 한 후 
이 다음에 필요한 데이터를 또 미리 불러오는 방식으로 구현해보았습니다.
이런 아이디어에 대해 어떻게 생각하시는지에 대해 궁금하고, 날씨의 경우 현업에서 어떻게 해결하고 있는지에 대해
궁금합니다!!

- 무한 스크롤의 기능을 활용하여 스크롤 하단에 도달시 다음 페이지에 대한 데이터를 가져오는 기능을 구현하였습니다.
```swift
extension ItemListViewController: UICollectionViewDelegate {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        guard !isFeching else {
            return
        }
        
        let position = scrollView.contentOffset.y
        if position > (marketItemListCollectionView.contentSize.height - scrollView.frame.height) {
            itemList += nextItemList
            nextItemList = []
            marketItemListCollectionView.reloadData()
            fetchNextItemList()
        }
    }
}
```
기본 로직은 조건 만족시 미리 준비해둔 데이터를 현재 데이터와 합친 후 리로딩을 하는데
같은 데이터가 혹시라도 반복되어 리로딩되는 경우를 방지하기위해 리로딩 전에 빈 배열로
변경하는 코드가 추가되어있는 상태입니다.

YongHoonJJo committed:
refactor: 테스트 코드를 위한 소스파일 타겟 위치 변경

- MockApiClient.swift, *.json 파일들, 테스트 타겟으로 위치 이동

jaemuYeo committed:
refactor: 모델 타입 분리 및 네이밍 수정

- Request 메소드별 타입 분리
- Request와 Response 구분에 따른 폴더 분리

YongHoonJJo committed:
feat: CollectionView 및 기본 셀 구성

- VieController 네이밍 수정 (ItemListViewController로 변경)
- ItemListViewController 를 collectionView 의 dataSource 및 delegate 로 연결(스토리보드상에서 진행)
- ItemListViewController에 네이게이션 컨트롤러 연결
  > 타이틀 설정, Bar Button Item 추가
- collectionView 에 대한 IBOutlet 연결
- UICollectionViewDataSource 프로토콜 채택 및 준수
- Custom Cell 생성(ItemCollectionViewCell) 및 UI 구현(필요한 IBOutlet 연결)
- 스토리보드상에서 셀에 대한 오토레이아웃 적용

jaemuYeo committed:
feat: MarketPageItem Api 호출 및 데이터 바인딩

- ApiClient 인스턴스 생성
- [MarketPageItem] 타입 프로퍼티 추가
- fetchItemList() 메서드 구현
- configureLabels(on:)메서드를 통해 셀에 데이터 바인딩

YongHoonJJo committed:
feat: Cell image Configure 및 Layout Design.

- updateThumbnailImage(to:) 를 통해 셀에 섬네일 적용
- configure(with:) 메서드를 통해 레이블 및 이미지 통합 업데이트하는 메서드 추가.
- cell 에서 price 관련 정보에 CURRENCY 추가.
- 스토리보드에서 ImageView 에 width, height 에 대한 제약 추가
- cell 에 대한 flowlayout 적용
  > 디바이스 너비에 따른 카드 너비 계산
- cell 테투리 디자인(cell.layer)

jaemuYeo committed:
feat: fetchItemList에 대한 무한 스크롤 적용

- UICollectionViewDelegate 채택
- scrollViewDidScroll(_ scrollView:) 준수
- 디바이스 테스트를 위한 Development Team, Project Bundle Identifier 변경

YongHoonJJo committed:
feat: Label 들에 대한 조건에 따른 스타일 적용

- cell.layer border coloer 변경
- 스토리보트에서 타이틀 레이블 height 에 대한 제약 추가, Label 들 align center 적용

jaemuYeo committed:
feat: Image 캐싱을 통한 최적화

- ImageManager 싱글턴 구현
- cacheImageData(imageUrl:) - 이미지를 데이터로 캐싱하는 메서드
- cacheImageData(imageUrl:data:) - 이미지를 데이터로 캐싱하는 메서드
- loadImageData(imageUrl:) - 캐싱한 데이터를 로드하는 메서드

YongHoonJJo committed:
feat: launchScreen 추가 및 첫 페이지 데이터 fetch 시 인디케이터 적용

jaemuYeo committed:
feat: 스크롤 기능 최적화

- fetchNextItemList() - 다음 페이지에 대한 데이터를 미리 준비해 최적화 적용 구현

YongHoonJJo committed:
refactor: 리팩토링

- 중복코드 제거
- 이미지 없는 경우, 기본 이미지로 띄어지도록 적용
- 접근제어 추가
- 폴더 구조 조정

YongHoonJJo commented:
### 추가 질문
```swift
DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
    self.isFeching = false
    self.fetchNextItemList()
}
```

에서 0.5초 뒤에 실행을 시킨 이유는 스크롤 이벤트가 매우 빠르게 발생하여, isFetching 으로 가드를 해도 다음 페이지에 대한 데이를 불러올 때, 한페이지씩이 아닌 3-4 페이지씩 불러오는 현상이 있었습니다. 이러한 현상을 방지하고자 위와 같이 조작을 하게 되었습니다. 
좋지 않은 방법이라면 대안 혹은 대안이 될 수 있는 레퍼런스 부탁드리겠습니다!

SungPyo commented on OpenMarket/OpenMarket/Managers/ImageManager.swift:
```diff
@@ -0,0 +1,46 @@
+import Foundation
+
+class ImageManager {
+    private var imageDataMap: [String: Data] = [:]
```
Dictionary 를 사용하신 이유가 있을까요?
NSCache 와의 차이점을 알고 계실까요?
만약 바로 답변이 어렵다면 문서를 한번 살펴봐 주세요!

SungPyo commented on OpenMarket/OpenMarket/Managers/ImageManager.swift:
```diff
@@ -0,0 +1,46 @@
+    
+    private init() {
+        guard let url = URL(string: noImageDataUrl),
+              let data = try? Data(contentsOf: url) else {
```
네트워크에 요청하는 imageUrl 입니다.
Data(contentsOf:~ 에 대한 애플 공식 가이드 문서 정독을 다시한번 확인 요청합니다.
만약 읽어도 이유를 모르시겠다면 디코에 남겨주세요!

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+
+class ItemListViewController: UIViewController {
+    private let apiClient = ApiClient()
+    private var itemList: [MarketPageItem] = []
```
간단하게 생각하면 모델을 가지고 화면을 그리면 됩니다. 
하지만 모델이 두개군요.  두개인 이유가 있을까요?
 첫 로드와 이후 로드를 분리하기 위해 나누신것 같다고 판단되는데, 왜 두개로 하셨는지 정확한 이유가 궁금합니다.

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+    private var itemList: [MarketPageItem] = []
+    private var nextItemList: [MarketPageItem] = []
+    private var nextPage = 1
+    private var isFeching = false
```
코드를 봤을때 isfeching 의 정확한 의미를 알 수 없습니다. 불러오기 중이다 라는걸 알기 위해서 쓰는걸까요?

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+        fetchItemList()
+    }
+    
+    private func fetchItemList() {
```
대충보면 fetchNextItemList 함수와 90%이상이 동일한 중복 코드가 되는것 같습니다.
최초 로드와 추가 로드를 구분하려고 두개로 나누신것 같다고 판단됩니다.
위에 코멘트를 달았지만, 모델을 두개로 나눈것도 이때문으로 파악이 되는데
요청 후 모델을 주입하기만 하는 역할만 한다면 두개가 나뉘어질 필요는 없어 보입니다. 혹시 다른 이유가 있을까요?

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+            case .success(let marketPageItem):
+                if marketPageItem.items.count > 0 {
+                    for item in marketPageItem.items {
+                        ImageManager.shared.performBatchUpdate(urls: item.thumbnails)
```
가져온 모든 아이템에 대한 섬네일 이미지의 url을 구성하는 함수로 보여집니다. 
 예를들어 한 페이지에 1000장씩 있다고 가정했을때 1000장의 모든 썸네일 이미지를 구성하는 함수입니다. 
 예시는 극단적이지만 화면에 당장 보여지는 사진이 아님에도 불필요한 자원을 사용하고 있다는 느낌이 강하게 듭니다. 
 보여지는 부분만 즉각 즉각 보여주기 위해서 어떻게 해야할지 고민이 필요할 것 같습니다. 
 (셀의 더보기 수정 관련 코멘트와 연결되는 부분입니다.)

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+                    self.nextPage = marketPageItem.page + 1
+                    
+                    DispatchQueue.main.async {
+                        self.loadingIndicator.stopAnimating()
```
stop 은 있는데 start가 없는 인디케이터네요. 초기값이 start 인거 같은데, 그럼 start를 해야하는 경우엔 어떻게 써야하나요?

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+                self.handleError(error)
+            }
+            
+            DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
+                self.isFeching = false
+                self.fetchNextItemList()
```
이러한 조작을 가하더라도 질문하신 내용에 완벽한 가드를 할 수 없습니다.
collectionView의 더보기 구성(prefetch, willDisplay등.. ) 을 한번 더 살펴봐 주세요.

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+    
+    private func fetchNextItemList() {
```
위 함수와 비슷한 부분이 많아 위 함수의 질문으로 대체하겠습니다.

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+        }
+        
+        let marketItem = itemList[indexPath.item]
+        cell.configure(with: marketItem)
+        
+        cell.layer.borderWidth = 1
+        cell.layer.borderColor = UIColor.lightGray.cgColor
+        cell.layer.cornerRadius = 8
```
cell을 구성하는 configure를 만들어놓고 layer는 또 이곳에 구성하셨네요?!
특별한 이유가 있을까요? 
특별한 이유가 없다면 cell이 초기화 되는 시점의 라이프 사이클을 한번 더 찾아보시는걸 추천드립니다.

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+}
+
+extension ItemListViewController: UICollectionViewDelegate {
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        guard !isFeching else {
+            return
+        }
+        
+        let position = scrollView.contentOffset.y
+        if position > (marketItemListCollectionView.contentSize.height - scrollView.frame.height) {
+            itemList += nextItemList
+            nextItemList = []
+            marketItemListCollectionView.reloadData()
+            fetchNextItemList()
+        }
+    }
```
요 함수 내부에 print를 찍어보면 보셨겠지만 불필요한 동작이 일어납니다.
또한 위 요청 함수에서 지정한 isFeching 플래그가 변경되는 시점에 하단의 로직이 불릴 가능성도 있습니다.
(위 코멘트로 달았지만 afterAsync 같은 플래그 조작으로 완벽하게 막아지지 않는다는 소리와 같습니다.)
collectionView에서 더보기를 구성(prefetch, willDisplay등.. ) 하는 방법에 대해서 한번 더 찾아볼 필요가 있어보입니다.

SungPyo reviewed (changes requested):
안녕하세요! jamking, jost

Step2 리뷰에 대해서는 전체 리뷰를 한번에 하는건 무리가 있어 보이네요.

부분 리뷰 후 수정 결과에 대해서 확인이 더 필요할 것 같아 각 코드별로 중요한 부분들만 추려서 코멘트를 달았습니다. 1차적으로 수정 반영을 한번 고려해 주세요.

### 지난 피드백 반영 내용

테스트 데이터 분리는 확인했습니다. 하지만 Request 모델에 대해서는 파일별 분리는 되어있으나, 사용하는데에 해당 모델을 사용하지 않는걸 확인했습니다. 따로 코멘트를 달진 않았지만 코드를 수정하시면서 어떻게 사용해야 할지 한번쯤은 확인할 필요가 있어 보입니다. (ex: 모델을 인자로 받는다던지.. 또는 타입을 인자로 받는다던지..)

### 매직넘버 관련

매직넘버를 지양하도록 말씀 드린건 사실이지만 해당 부분처럼 어쩔 수 없는 계산식이 들어가는 경우에는 사용할 수 밖에 없겠죠?! 해당 값을 여러곳에서 사용한다면 전역변수로 지정하여 사용하는걸 추천합니다!

레이아웃에 대해서는 크게 확인하지 않았습니다. 말씀하신것 처럼 제가 경험한 실제 프로젝트에서는 디자인의 요소는 디자이너가 지정해주는 크기나 비율에 따라 그대로 적용되기 때문에 이 부분은 자유롭게 진행하셔도 좋을 것 같아요. 세로 길이에 대해서는 estimatedSize 에 대해서 한번 찾아보시는걸 추천드립니다.

### 최적화 관련

최적화 관련(캐싱 등..) 에 대해서는 많은 생각을 하셨을 것이고, 고민하여 개발하셨을 테지만 제가 본 현재 코드는 최적화와 거리가 좀 멀다고 생각됩니다. 
쓴소리를 조금 더하자면 첫 요청 부터 셀을 불러오는 방법, 이미지 캐싱 방법 등 어느 하나도 제 기준의 50%도 맞춰지지 않았다고 생각됩니다. 그 이유로는 중복 호출이나 이상동작을 완벽하게 막을 수 없는 무분별한 플래그 사용(플래그 사용 자체가 잘못 되었다는건 아닙니다.), NSCache를 사용하지 않은 문제, 네트워크 기반의 Image Data 호출 방법, 불필요한 메모리 사용, 더보기 구현 방법의 문제 등 여러 문제들이 있기 때문입니다. 이 프로젝트(Step2) 에서 개인적으로 제일 중요하다고 여겨지는 부분은 **이미지 캐시 구성**과 **모델 구성을 포함한 더보기 구성** 방법입니다.

하지만 코드를 보지 않고 그냥 해피케이스의 상태로만 동작을 봤을때 올바르게 동작하고(최고 칭찬👍) , 문제를 해결하려는 노력은 계속 봐 왔기때문에 그 부분은 칭찬해 드리고 싶습니다.

코멘트를 한번 살펴 보시고, 한번 더 확인해 주세요!

이해가 가지 않는 부분이 있다면 질문 해 주세요! 
다만 이번주 주말에는(일요일 제외) 응답이 조금 힘들 수 있어요! 이해가 되지 않는 부분을 알려주시면 확인되는대로 응답 해 드릴게요!

화이팅입니다.

jaemuYeo committed:
refactor: 불필요한 프로퍼티와 메서드 제거 및 셀 디자인 코드 위치 변경

- nextItemList, isFeching 프로퍼티 제거 및 적용 부분 수정
- fetchNextItemList() 메서드 제거
- 셀 스타일 관련 코드 awakeFromNib() 메서드를 통해 한번만 호출되도록 변경
- scrollViewDidScroll(_:) 메서드 제거

YongHoonJJo committed:
feat: ImageCacheManager 타입 선언 및 프로퍼티, 메서드 구현

- ImageCacheManager: 싱글턴 패턴으로 구현
- 이미지 캐싱을 위한 NSCache 적용
- ItemCollectionVIewCell 에서 추후 NSCache 적용을 위해 updateThumbnailImage(to:) 메서드 삭제 및 updateThumbnail(to:) 추가

jaemuYeo committed:
feat: ImageDownloadManager 타입 선언 및 프로퍼티, 메서드 구현

- Data(contentsOf:) 메서드를 대체하기 위한 urlSession을 이용한 이미지 다운로드 로직 구현

YongHoonJJo committed:
feat: 셀에 이미지 바인딩

- 캐시된 이미지가 있을경우 적용.
- 없을 경우 URLSession 을 이용한 이미지 다운로드 및 적용, 캐싱 처리
- reloadCollectionView(with indexPath:) : indexPath 에 해당하는 셀이 화면에 표시된 경우 UI 업데이트 처리.

jaemuYeo committed:
feat: 스크롤성능 향상을 위한 최적화, 페이지네이션 적용

- UICollectionViewDelegate, UICollectionViewDataSourcePrefetching 딜리게이트 활용
- collectionView(_:willDisplay) -   셀이 화면에 표시되기 직전에 데이터를 한번 더 확인하는 로직 추가
- collectionView(_:prefetchItemsAt:) - data prefetching 구현

YongHoonJJo committed:
refactor: flowLayout 을 DelegateFlowLayout 을 통해 적용

- 스토리보드에서 collectionView 의 Estimate Size 를 None 으로 변경.
- 뷰컨트롤러를 prefetchDataSource 로 지정하는 코드 누락으로 인한 적용.

jaemuYeo commented:
@SungPyo 
날씨 안녕하세요!! Step2에 대한 조언들을 반영하여 리팩토링을 진행하였습니다!
이번 리뷰도 잘 부탁드리겠습니다 🙏

### 피드백 반영 내용
- 네트워크에 요청하는 imageUrl에 대해 Data(contentsOf:)를 사용한 부분을
URLSession을 통해 적용하도록 변경하였습니다.
Data(contentsOf:)는 동기적으로 작동하기 때문에 네트워크를 Request에 활용하는 것은
적합하지 않다는 것을 확인하였습니다.

- Dictionary를 통한 이미지 캐싱을 NSCache로 대체하였습니다.
스레드 safe와 메모리를 자동으로 관리해주는 것이 큰 장점인 것을 숙지하였습니다.

- Cell prefetching과 Data prefetching에 대한 부분을 확인 후 적용하였고,
이로인해 기존의 flag 사용과 fetchNextItemList() 메서드등이 불필요한 코드였다는 것을 알게되었습니다.
또한 scrollViewDidScroll()메서드를 통한 로직도 불필요한 코드였던 것 같습니다.

- UICollectionViewDataSource에서 Cell을 재사용 할 때마다 레이아웃을 적용하던 부분 역시 
불필요하게 반복되고 있어서,  Cell이 처음 생성당시 한번만 디자인하면 된다고 판단하여
awakeFromNib()을 통해 최초의 Cell에 디자인되도록 반영하였습니다.

### 조언 받고 싶은 부분
- FlowLayout을 이전 PR에서는 객체를 생성하여 viewDidLoad()를 통하여 레이아웃이
반영하도록 하였고, 이번 PR에서는 Delegate를 통해 반영할 수 있는 방법을 알게되어 공부를 하고 적용해보았습니다.
두 가지 방식에 어떤 차이점이 있을까 고민을 해 보았는데, 
 기존 방식은 레이아웃이 변경이 되지 않을 경우 적합한 모델일 것이라 생각을 하였고,
이번에 적용한 방식은 그리드의 모양을 전환할 필요가 있을 경우 적합할 것이라고 예상했는데
이러한 예상이 적절한지 궁금합니다.

- 무한스크롤 페이징 관련하여 preloadCount 프로퍼티에 8이라는 값을 할당했습니다.
이는 데이터 프리패칭시 데이터가 8개가 남았을 때, 다음 페이지에 대한 데이터를 불러오는
기준값에 해당합니다. 디바이스로 테스트해본 결과 페이지가 로딩하는 순간에 살짝의
로딩지연이 있었습니다.
혹시 페이지 로딩을 할 때 부드럽게 데이터를 불러올 수 있는 값 또는 다른 방식에 대해
어떻게 구현하면 좋을지 궁금합니다.

- NSCache 관련하여 NSString으로 셋팅하였는데, 적절한 방법인지 궁금합니다.
적절한 방법인지 여쭤보는 이유는 썸네일 이미지 주소가 긴 String에 해당하기 때문에 그렇습니다.
또한 NSString이전에 NSNumber를 키값으로 설정하는 방법에 대해 고민해 보았고,
이 키 값에는 Item의 id를 적용하는 방법을 먼저 생각해 보았습니다.
하지만 이 경우 Item의 썸네일이 변경이 되었을 때 이를 대응하지 못할 것이라고 판단을 하였는데
적절한 선택이였는지, 다른 대안이 있어서 NSNumber를 키 값으로 적용하는 방법이 맞는지에 대해
조언하고 싶습니다.

SungPyo reviewed (changes requested):
Jost, Jamking 
주말동안 수정하시느라 고생하셨습니다.

괜찮게 수정된 부분도 있고 아닌 부분도 존재합니다. 

그래도 내용을 보니 많은걸 공부하고 적용한 듯 싶네요! 👍 

조금 냉정하게 보일 수 있겠지만 저의 리뷰 결과는 이전과 크게 다르지 않습니다.

우선 로직 자체가 눈에 들어오지 않고, 너무 복잡합니다.

분기도 많구요.

제가 생각한 해피케이스를 말로 한번 간단하게 설명해 볼게요. 

다음 수정시 참고가 되었으면 좋겠습니다.

1. 화면이 나타나면
2. 데이터를 요청하고
3. 데이터를 받아오면 모델을 구성한다.
4. 모델이 변경되면
5. 가지고있는 모델을 기반으로 뷰를 리로드 한다.
6. 셀이 그려질때, 가지고 있는 모델로 셀을 구성한다
7. 셀 구성시 모델 내부의 Url을 이용해 이미지를 요청한다 
7-1. 캐시에 존재한다면 요청하지 않는다.
7-2. 이미지를 받아왔다면 캐시에 저장한다.
8. 받아온 이미지를 셀 내의 이미지 뷰에 넣어준다
9. 스크롤을 내려 끝 index가 보이면 -> `2`로 간다
10. 셀이 재사용 될때 요청 Task와 이미지를 제거한다.

자 이제 두분의 로직중 의아한 부분만 몇군데 써볼까요?

1. 셀이 그려지려고 할때 해당 cell이 ItemCollectionViewCell 이면 해당 cell을 구성한다
2. 셀이 그려질때 ItemCollectionViewCell을 가져와서 셀을 구성한다. 

요 두가지만 보더라도 로직이 중복될 수 있겠죠? `(will~, prefetch 를 둘 다 사용할 필요가 없습니다 개발자 문서 한번 더 정독!)`

이미지가 없으면 downloadImage 를 통해서 새로 요청하는데, 셀이 나타나기 직전, 나타날때, prefetching 시 모두 호출하다보니 불필요한 요청이 생깁니다.

downloadImage 내부의 가드문 또한 성격이 다른 조건들을 한곳에 우겨넣다 보니 가독성이 굉장히 떨어집니다. guard 문 내부의 contains클로저는 사실 보기에 굉장히 불편해서 읽고싶지 않을 정도입니다. `가드문이나 if문을 여러개 사용하더라도 동일한 성격에 맞게 분리하는걸 추천합니다`

URLSession의 기본 사용법이나, NSCache의 기본 사용법은 이해하신대로 잘 사용하고 계신듯 합니다.  추가 옵션들도 있으니 이 부분은 나중에 한번 보시면 좋을 것 같아요 
(요 프로젝트에선 없어도 무관할 것이라 판단됩니다.)

completeTask 함수의 용도는 사실... 잘 모르겠습니다. 의도가 궁금합니다.

이미지 요청 후 유효성 검사는 해당 index값으로 하는것으로 보여지는데, prefetchItemsAt 을 통해 들어오는 downloadImage 에서 중복된 index를 무작위로 요청하는 경우가 생길 수 있을 것 같습니다. index는 여러번 불릴 수 있고, 로직상 중복호출이 될 수 있는 위험도 충분히 있기에 모델을 통한 셀 구성을 제외한 indexPath의 사용은 권장하지 않습니다. 이미지 요청시 `url은 모델별로 고유하거나 같으므로 url만 사용하셔도 충분히 구성할 수 있습니다.`
downloadImage 의 함수 또한 필요한 부분에만 불려야 맞지만 화면이 그려지는 내내 그려진다고 보여집니다.

위에 작성한 해피케이스를 한번 더 살펴보시고 불필요한 indexPath의 구성과 이미지 구성에 대해서 다시한번 수정되었으면 좋겠습니다. 
셀 재사용에 관련되어서는 `prepareForReuse` 에 대한 학습이 필요하니, 이 부분도 같이 학습 해 보셨으면 합니다.

화이팅!

SungPyo commented:
> @SungPyo
> 날씨 안녕하세요!! Step2에 대한 조언들을 반영하여 리팩토링을 진행하였습니다!
> 이번 리뷰도 잘 부탁드리겠습니다 🙏
> 
> ### 피드백 반영 내용
> * 네트워크에 요청하는 imageUrl에 대해 Data(contentsOf:)를 사용한 부분을
>   URLSession을 통해 적용하도록 변경하였습니다.
>   Data(contentsOf:)는 동기적으로 작동하기 때문에 네트워크를 Request에 활용하는 것은
>   적합하지 않다는 것을 확인하였습니다.
> * Dictionary를 통한 이미지 캐싱을 NSCache로 대체하였습니다.
>   스레드 safe와 메모리를 자동으로 관리해주는 것이 큰 장점인 것을 숙지하였습니다.
> * Cell prefetching과 Data prefetching에 대한 부분을 확인 후 적용하였고,
>   이로인해 기존의 flag 사용과 fetchNextItemList() 메서드등이 불필요한 코드였다는 것을 알게되었습니다.
>   또한 scrollViewDidScroll()메서드를 통한 로직도 불필요한 코드였던 것 같습니다.
> * UICollectionViewDataSource에서 Cell을 재사용 할 때마다 레이아웃을 적용하던 부분 역시
>   불필요하게 반복되고 있어서,  Cell이 처음 생성당시 한번만 디자인하면 된다고 판단하여
>   awakeFromNib()을 통해 최초의 Cell에 디자인되도록 반영하였습니다.
> 
> ### 조언 받고 싶은 부분
> * FlowLayout을 이전 PR에서는 객체를 생성하여 viewDidLoad()를 통하여 레이아웃이
>   반영하도록 하였고, 이번 PR에서는 Delegate를 통해 반영할 수 있는 방법을 알게되어 공부를 하고 적용해보았습니다.
>   두 가지 방식에 어떤 차이점이 있을까 고민을 해 보았는데,
>   기존 방식은 레이아웃이 변경이 되지 않을 경우 적합한 모델일 것이라 생각을 하였고,
>   이번에 적용한 방식은 그리드의 모양을 전환할 필요가 있을 경우 적합할 것이라고 예상했는데
>   이러한 예상이 적절한지 궁금합니다.
> * 무한스크롤 페이징 관련하여 preloadCount 프로퍼티에 8이라는 값을 할당했습니다.
>   이는 데이터 프리패칭시 데이터가 8개가 남았을 때, 다음 페이지에 대한 데이터를 불러오는
>   기준값에 해당합니다. 디바이스로 테스트해본 결과 페이지가 로딩하는 순간에 살짝의
>   로딩지연이 있었습니다.
>   혹시 페이지 로딩을 할 때 부드럽게 데이터를 불러올 수 있는 값 또는 다른 방식에 대해
>   어떻게 구현하면 좋을지 궁금합니다.
> * NSCache 관련하여 NSString으로 셋팅하였는데, 적절한 방법인지 궁금합니다.
>   적절한 방법인지 여쭤보는 이유는 썸네일 이미지 주소가 긴 String에 해당하기 때문에 그렇습니다.
>   또한 NSString이전에 NSNumber를 키값으로 설정하는 방법에 대해 고민해 보았고,
>   이 키 값에는 Item의 id를 적용하는 방법을 먼저 생각해 보았습니다.
>   하지만 이 경우 Item의 썸네일이 변경이 되었을 때 이를 대응하지 못할 것이라고 판단을 하였는데
>   적절한 선택이였는지, 다른 대안이 있어서 NSNumber를 키 값으로 적용하는 방법이 맞는지에 대해
>   조언하고 싶습니다.

1. `플로우 레이아웃과 델리게이트 구현의 사이에서 망설이신다면! 원하는대로 그려진다면 무관하다! 라고 답변해 드리고 싶습니다!`
2. `프리패칭 관련하여 프리패칭이 어떻게 동작하는지 한번 더 살펴보셔야 할 것 같습니다. 무작정 리로드 하는게 아니라 프리패칭에서 내려오는 **indexPath및 가지고 있는 모델의 배열을 이용하여 마지막 indexPath가 있는 경우 리로드를 하도록 구성** 하시면 불필요한 호출없이 부드럽게 호출됩니다. 이게 무조건 맞는 방법은 아니니 답변을 참고하여 다른 방법을 생각해 보시는 것도 충분히 도움이 됩니다.`
3. `url은 이미지의 유일한 키값이고, 기본키가 될 수 있습니다. 문제가 없다면 NSNumber를 사용하셔도 무관하지만, NSString(url) 을 추천합니다. 물론 같은 url이지만 이미지가 변경되는 스펙이 있는 경우 etag라는걸 이용하여 구분하는 방법이 있지만 현재 단계에서는 그 부분까지 학습할 필요는 없기에 패스할게요!`

YongHoonJJo committed:
refactor: 가드문 분리.

YongHoonJJo committed:
refactor: 피드백 반영

- ImageDownLoadManager 의 downloadImage() 메서드에서 불필요한로직 삭제
- ImageDownLoadManager 의 downloadImage() 메서드에서 불필요한 매개변수 삭제
- ImageDownLoadManager 의 downloadImage() 메서드에서 completionhandler 관련 로직 수정
- ImageDownLoadManager 의 downloadImage() 메서드 static 으로 변경

- ItemCollectionViewCell 에서 prepareForReuse() 메서드 추가.
- ItemCollectionViewCell 내에서 이미지 캐싱 및 다운로드, 업데이트 로직 적용

- ItemListViewController 에서 불필요해진 메서드들 삭제
- ItemListViewController 의 collectionView(_ collectionView:prefetchItemsAt:) 메서드 내부 로직 변경

jaemuYeo committed:
refactor: 불필요한 파일 제거

- updateThumbnail(to:) - private 적용
- prepareForReuse() - 이미지만 초기화

jaemuYeo commented:
@SungPyo 
날씨 안녕하세요 ㅎㅎ 다시 리뷰 부탁드립니다 😀
참고로 주신 해피케이스와 피드백을 반영하여 다시 한번 수정해보았습니다.
이번 리뷰도 잘 부탁드립니다!!

### 피드백 반영 내용
- guard문 분리
- 이미지 요청 시 url로만 요청하도록 적용
- Delegate의 willDisPlay 제거
- Cell 내에서 prepareForReuse() 사용 및 이미지 요청 로직 적용
- 데이터 프리패칭 시 마지막 인덱스가 보이면 다음 페이지 데이터를 요청하도록 적용

### 조언받고 싶은 부분
- prepareForReuse()를 통해 이미지를 초기화 하였습니다.
다른 컨텐츠들은 분기처리를 하여서 nil로 초기화 하지않고, 이미지만 하였는데
잘 적용한건지 궁금합니다.

SungPyo reviewed (changes requested):
조금씩이지만 코드가 점점 좋은 방향으로 변화하고 있네요. 

셀 재사용시 이미지만 초기화 하는게 옳은가 에 대한 질문을 해주셨는데요.

셀이 재사용되거나 새로 그려질때, 즉시 반영되지 않고, 비동기적으로 변화하거나, 적용되는 속성은 재사용시 초기값을 지정해 주어야 UI가 부드럽게 동작합니다. 이미지도 그중 하나지요 :) 이거로 답변이 되었을까요?

코드 내용에 대해서도 아직 문제가 존재하고 있습니다.

해당 부분에 대해서 코멘트를 새로 달아놓았으니, 참고하셔서 수정해 보세요! 

그 외에 코멘트에 저의 질문도 하나 추가로 달았습니다 :)

마지막까지 화이팅입니다!

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,127 @@
+                ImageDownloadManager.downloadImage(with: thumbnailUrl)
+            }
+            
+            if indexPath.item == lastitem {
+                fetchItemList()
+            }
```
이러한 조건이라면 이미 다 받아온 경우에도 스크롤을 올렸다 내렸다 한다면, 불필요한 요청이 지속될 것이라고 생각됩니다. 
어떠한 방법을 사용해야 불필요한 요청을 최소화 할 수 있는지도 한번 생각해 주세요!

YongHoonJJo replied:
고민고민 끝에 가장 적합하다고 생각하는 곳으로 코드 위치를 옮겼는데요, 의도하신 바가 맞는지 궁금합니다!

SungPyo replied:
아뇨 제가 의도한 바와는 조금 다른 것 같습니다.
제가 의도한 부분은 불러오기를 준비해야 하는 상태에서 불러오기 '만' 하는것을 희망했습니다.
하지만 현재 동작은 해당 위치에 이미지가 있는지도 확인하고 들어오는 모든 indexPath에 대해서 이미지가 있는지 없는지까지 검사하고 있죠. 저는 이런 부분이 불필요하다고 생각됩니다.
prefetchItemsAt 메서드는 찾아보면서 공부하셨겠지만 
해당 indexPath를 로드할 준비가 되면 불리는 함수입니다. 셀을 그리기 직전에 불리는 함수가 아니라는거죠.
짜여진 코드를 보면 메서드에 대해 어느정도 공부 한 것으로 보여집니다.
글로 전달이 쉽지 않은 것 같고 시간도 충분하지 않아 보충할만한 부분은 디코에서 추가적으로 보충하는 것으로 할게요.

저의 기준에서 Step2에서 중요하다고 생각되는 부분들에 대해서 어느정도 설명이 됐다고 생각하여

별도의 코멘트 없이 리뷰 마치겠습니다 :)

3주간 고생 많으셨습니다.

SungPyo commented on OpenMarket/OpenMarket/ViewControllers/ItemListViewController.swift:
```diff
@@ -0,0 +1,127 @@
+extension ItemListViewController: UICollectionViewDataSourcePrefetching {
+    func collectionView(_ collectionView: UICollectionView, prefetchItemsAt indexPaths: [IndexPath]) {
+        for indexPath in indexPaths {
+            let marketItem = itemList[indexPath.item]
+            let thumbnailUrl = marketItem.thumbnails[0]
+            let lastitem = itemList.count - 1
+            
+            if ImageCacheManager.shared.loadCachedData(for: thumbnailUrl) == nil {
+                ImageDownloadManager.downloadImage(with: thumbnailUrl)
+            }
```
[질문] ]marketItem을 뽑아서 이미지를 다운로드 하는 이유가 뭔가요? 어떤 이유가 있는건지 궁금합니다.

YongHoonJJo replied:
예전에 코드를 작성하는 과정에서 비슷한 로직은 타이핑하지 않고 복붙해와서 해당 메서드의 내용에 맞게 수정하는 식으로 코드를 작성하였는데, 그런 과정에서 질문하신 부분은 굳이 따로 상수에 할당을 하지 않아도 되는데, 코드를 작성할 당시에 크게 신경을 쓰지 못했던 것 같습니다. ㅜ

SungPyo replied:
아하.. 레거시로군요

SungPyo commented on OpenMarket/OpenMarket/Views/ItemCollectionViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+            itemThumbnailImageView.image = image
+        }
+        else {
+            ImageDownloadManager.downloadImage(with: thumbnailUrl) { image in
+                self.itemThumbnailImageView.image = image
+            }
```
이미지를 요청하고 응답을 받는 사이 셀이 재사용이 되어버리면 이미지를 셋팅하는데 문제가 없을까요? 
만약 문제가 생긴다면 어떤 문제가 생길것 같은지 한번 생각하셔서 코멘트 달아주셨으면 좋겠습니다!

YongHoonJJo replied:
일단 준비된 이미지가 업데이트 될 당시의 셀의 이미지가 아닐 경우가 발생할 수 있는데 그럼에도 불구하고 해당 셀에 업데이트를 하는 로직이 될 것 같습니다.
`이미지 요청시 url 은 모델별로 고유하거나 같으므로 url 만 사용하셔도 충분히 구성할 수 있습니다.` 라고 말씀해주신 부분은 이제서야 이해했습니다. :D

SungPyo commented on OpenMarket/OpenMarket/Views/ItemCollectionViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+    }
+    
+    private func updateThumbnail(to marketItem: MarketPageItem) {
+        let thumbnailUrl = marketItem.thumbnails[0]
```
섬네일이 만약 하나도 없는 경우가 있다면 크래시가 발생할 것 같다고 생각되는데, 더 좋은방법 없을까요?

YongHoonJJo replied:
<https://developer.apple.com/documentation/swift/array/1689165-first> 를 활용하였습니다!

YongHoonJJo committed:
refactor: 프리패칭시 fetchItemList() 의 호출 최소화 문제 개선.

YongHoonJJo committed:
refactor: 섬네일 데이터가 없는 경우에 대한 예외처리.

YongHoonJJo committed:
refactor: 불필요하게 상수에 값 할당하는 로직 수정.

YongHoonJJo committed:
feat: 이미지 요청시 url 의 검증을 통한 이미지 업데이트 적용되도록 로직 수정.

jaemuYeo committed:
feat: 셀이 재사용 될때 요청 Task와 이미지 제거

- 해피케이스 10번 반영

SungPyo reviewed (approved):
별도의 코멘트 없이 기존 코멘트로 마무리 할게요 :)

3주간 고생 많으셨습니다.
